### PR TITLE
Refactoring of getting balance

### PR DESCRIPTION
**Pull request by prazd**

### CODE REVIEW

File: utils-for-endpoints/estorage/storage.go [hunk 1]
```diff
@@ -4,71 +4,178 @@ import (
...
-	case "etc":
-		s.EtcEndpoints = entry
-	}
+	s.Entry = entry
 	s.Unlock()
 }
 
-func (s *StoredEndpoints) Get(currency string) *schema.Endpoints {
+func (s *StoredEndpoints) Get() *schema.EndpointsData {
 	s.RLock()
```
Comment (krboktv) [inline utils-for-endpoints/estorage/storage.go]: Зачем тут RLock, если функция по дефолту на чтение и писать в нее никто не будет?

File: eth-based/main.go [hunk 2]
```diff
@@ -10,14 +10,15 @@ import (
 )
 
 func init() {
-	go estorage.StoreEndpoints()
+	go estorage.StoreEndpointsFromDB()
```
Comment (EnoRage) [inline eth-based/main.go]: estorage is not the best naming option

File: shared/balance/balance.go [hunk 3]
```diff
@@ -63,44 +63,69 @@ func GetUtxoBasedBalancesByList(addresses []string) (map[string]string, error) {
...
 
-	mainApi := os.Getenv("main-api")
+	mainApi := os.Getenv("MAIN_API")
 
 	mainUrl := mainApi + "/v1/address/" + address
 
 	switch currency {
 	case "btc":
-		dbEndpoints := estorage.EndpointsFromDB.BtcEndpoints.Addresses
+		dbEndpoints := estorage.EndpointsFromDB.Get().Addresses
```
Comment (EnoRage) [inline shared/balance/balance.go]: can be before switch

File: shared/balance/balance.go [hunk 4]
```diff
@@ -63,44 +63,69 @@ func GetUtxoBasedBalancesByList(addresses []string) (map[string]string, error) {
...
 		for _, j := range dbEndpoints {
 			j = j + "/addr/" + address
 			endpoints = append(endpoints, j)
 		}
 		endpoints = append(endpoints, mainUrl)
 	case "ltc":
-		dbEndpoints := estorage.EndpointsFromDB.LtcEndpoints.Addresses
+		dbEndpoints := estorage.EndpointsFromDB.Get().Addresses
```
Comment (EnoRage) [inline shared/balance/balance.go]: can be before switch

File: shared/balance/balance.go [hunk 5]
```diff
@@ -63,44 +63,69 @@ func GetUtxoBasedBalancesByList(addresses []string) (map[string]string, error) {
...
 		for _, j := range dbEndpoints {
 			j = j + "/api/addr/" + address
 			endpoints = append(endpoints, j)
 		}
 		endpoints = append(endpoints, mainUrl)
 	case "bch":
+		dbEndpoints := estorage.EndpointsFromDB.Get().Addresses
```
Comment (EnoRage) [inline shared/balance/balance.go]: can be before switch

File: shared/balance/balance.go [hunk 6]
```diff
@@ -134,43 +159,57 @@ func UtxoBasedBalanceReq(endpoints []string) (string, error) {
...
 	if err != nil {
 		return "", err
 	}
 
-	return balance, nil
+	res, err := instance.BalanceOf(nil, common.HexToAddress(userAddress))
+	if err != nil {
+		balance, err := TokenBalanceReq(userAddress, smartContractAddress)
+		if err != nil {
+			return "", err
```
Comment (EnoRage) [inline shared/balance/balance.go:199]: can we log errors to Elastic Search @maxUo what do you think?

File: shared/balance/balance.go [hunk 7]
```diff
@@ -204,7 +243,11 @@ func TokenBalanceReq(endpoints []string, userAddress, smartContractAddress strin
 	}
 }
 
-func EtherBalanceReq(endpoints []string, address string) (string, error) {
+func EtherBalanceReq(address string) (string, error) {
+
+	endpoints := estorage.EndpointsFromDB.Get().Addresses
+
+	endpoints = append(endpoints, os.Getenv("MAIN_API"))
```
Comment (krboktv) [inline shared/balance/balance.go:250]: For good, all you need to make env in config, I can tell you a package to give to him, not to climb on the entire project and not look for the ENV

File: utils-for-endpoints/estorage/storage.go [hunk 8]
```diff
@@ -4,71 +4,178 @@ import (
...
+		log.Println(runtime.NumGoroutine())
+
+		time.Sleep(time.Minute * 1)
+	}
+}
+
+func GetFastestUtxoBasedEndpoint(mainUrl string) string {
+	currency := os.Getenv("BLOCKCHAIN")
+
+	var endpoints []string
```
Comment (EnoRage) [inline utils-for-endpoints/estorage/storage.go]: add dbEndpoints := EndpointsFromDB.Get().Addresses here

File: utils-for-endpoints/esync/main.go [hunk 9]
```diff
@@ -35,6 +35,11 @@ type Req func(address, currency string) (int64, error)
...
 func SyncCheck(currency string, addresses []string) error {
 
+	if currency == "bch" {
+		log.Println(addresses)
+		return nil
+	}
+
 	var (
 		getBlockNumber  Req
 		blockDifference int64 = 5
```
Comment (EnoRage) [inline utils-for-endpoints/esync/main.go]: Maybe add delta to 10?

File: shared/db/db.go [hunk 10]
```diff
@@ -117,6 +118,26 @@ func GetAllNodes(currency string) ([]string, error) {
...
+		return nil, err
+	}
+
+	defer session.Close()
+
+	var entry schema.EndpointsData
+
+	c := session.DB(database).C(collection)
+
+	err = c.Find(bson.M{"currency": strings.ToLower(os.Getenv("BLOCKCHAIN"))}).One(&entry)
```
Comment (krboktv) [inline shared/db/db.go:133]: bson is better to make in the model, and accessed through a share. bson:"currency", for example

File: utils-for-endpoints/esync/main.go (hunk 9 above)
Comment (prazd) [reply utils-for-endpoints/esync/main.go]: okay

File: utils-for-endpoints/estorage/storage.go (hunk 8 above)
Comment (prazd) [reply utils-for-endpoints/estorage/storage.go]: Okay, thanks, didn't notice

File: shared/db/db.go [hunk 11]
```diff
@@ -117,6 +118,26 @@ func GetAllNodes(currency string) ([]string, error) {
...
+	session, err := mgo.DialWithInfo(&info)
+	if err != nil {
+		return nil, err
+	}
+
+	defer session.Close()
+
+	var entry schema.EndpointsData
+
+	c := session.DB(database).C(collection)
```
Comment (krboktv) [inline shared/db/db.go:131]: Another session (they are in each function), can be put in the repository

File: shared/balance/balance.go (hunk 6 above)
Comment (prazd) [reply shared/balance/balance.go:199]: I think we can tell about it on the next sprint

File: utils-for-endpoints/estorage/storage.go [hunk 12]
```diff
@@ -4,71 +4,178 @@ import (
...
+}
+
+func (f *FastestEndpoint) Set(addr string) {
+	f.Lock()
+	f.Address = addr
+	f.Unlock()
+}
+
+func (f *FastestEndpoint) Get() string {
+	f.RLock()
```
Comment (krboktv) [inline utils-for-endpoints/estorage/storage.go]: RLock зачем там, где только читаешь с переменной? Он блокирует запись на все горутины, но разрешает чтение. Так и так же считать можешь

File: utils-for-endpoints/estorage/storage.go [hunk 13]
```diff
@@ -4,71 +4,178 @@ import (
...
+
+func SetFastestEndpoint() {
+	log.Println("Started set fastest endpoint!")
+
+	var (
+		getEndpoint GetFastestEndpoint
+		mainUrl     string
+	)
+
+	switch os.Getenv("BLOCKCHAIN") {
```
Comment (krboktv) [inline utils-for-endpoints/estorage/storage.go]: again, all env should be a separate config and take from config

File: utils-for-endpoints/estorage/storage.go [hunk 14]
```diff
@@ -4,71 +4,178 @@ import (
...
-}
 
-func StoreEndpoints() {
-	log.Println("Start storing!")
 	for {
-		entries, err := db.GetAll()
-		if err != nil {
-			log.Println(err)
-			continue
+		result := getEndpoint(mainUrl)
```
Comment (krboktv) [inline utils-for-endpoints/estorage/storage.go]: Or maybe the Ticker to do it?

File: utils-for-endpoints/estorage/storage.go [hunk 15]
```diff
@@ -4,71 +4,178 @@ import (
...
+			_, err := ethClient.EthGetBalance(os.Getenv("ADDRESS"), "latest")
+			if err != nil {
+				return
+			}
+
+			fastestEndpoint <- e
+		}(e)
+	}
+
+	return <-fastestEndpoint
```
Comment (krboktv) [inline utils-for-endpoints/estorage/storage.go]: You can, in fact, to traverse up the overall channel and take one function of all channel values
Make the api and client

File: utils-for-endpoints/storage/storage.go [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+
+func (f *FastestEndpoint) Get() string {
+	f.RLock()
+	defer f.RUnlock()
+	return f.Address
+}
+
+type GetFastestEndpoint func(string) string
+
+var (
```
Comment (jeyldii) [inline utils-for-endpoints/storage/storage.go:53]: Instead 2 global var make 2 constructors and 2 interfaces like this:
```
type EndpointsStorage interface {
     //Your methods here
    //Get()
    //Set()
}

type FastestEndpointStorage interface {
     //Your methods here
    //Get()
    //Set()
}
```

```
func NewEndpointsStoreConstructor (EndpointsStorage) {
	return {
		mu: new(sync.Mutex),
		store: new(StoredEndpoints),
	}
}

func NewFastestEndpointStoreConstructor (FastestEndpointStorage) {
	return {
		mu: new(sync.Mutex),
		store: new(StoredEndpoints),
	}
}
```

So, in main.go you should do like this:

```
fastestStore := NewFastestEndpointStoreConstructor()
endPointsStore := NewEndpointsStoreConstructor()

// throw each of storage in every paths in your program, but it maybe need some refactors your project-model
//For example, you should throw it in your handlers:

// handlers.go
type controller struct {
  storage FastestEndpointStorage
}

func(c controller) SomeFunc(c *gin.Context) {
  // Get Request

 // You should answer with data from storage :
   fromStore := c.storage.Get()
   c.JSON(200, gin.H{"result": &fromStore}
}
```

Comment (prazd) [reply utils-for-endpoints/storage/storage.go:53]: Thank you, I'll think about it

File: utils-for-endpoints/estorage/storage.go (hunk 14 above)
Comment (prazd) [reply utils-for-endpoints/estorage/storage.go]: I think it's not matter

**Pull request merged**